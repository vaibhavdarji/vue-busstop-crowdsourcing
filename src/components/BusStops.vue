<template>
  <v-card :id="Constants.APP_TITLE.BS_LIST_ID">
    <v-container grid-list-xl fluid>
      <v-layout v-if="isLoading" row wrap justify-center align-center>
        <v-progress-circular indeterminate color="primary"></v-progress-circular>
      </v-layout>
      <template v-if="errorMsg">
        <custom-error :msg="errorMsg" @onErrorBtnClick="onRefresh"></custom-error>
      </template>

      <v-layout row wrap v-else>
        <v-flex xs4 v-for="stop in totalStops" :key="stop.stopId" :class="Constants.APP_BS_ITEMS.BS_CARD_CLS">
          <bus-stop :stop="stop"></bus-stop>
        </v-flex>
      </v-layout>
    </v-container>
  </v-card>
</template>

<script>
import { mapGetters, mapActions } from "vuex";
import Constants from '@/constants';
import CustomButton from "@/components/Button.vue";
import BusStop from "@/components/BusStop.vue";
import CustomError from "@/components/Error.vue";
import {
  GET_ALL_STOPS
} from '@/actions/types';
export default {
  name: "BusStops",
  data() {
    return {
      errorMsg: "",
      doRefresh: false
    };
  },

  components: {
    BusStop,
    CustomButton,
    CustomError
  },

  asyncComputed: {
    totalStops: {
      get() {
        return this.busStops.length
          ? this.busStops
          : this.getAllStops()
              .then(data => {
                this.doRefresh = false;
                return data;
              })
              .catch(error => {
                this.errorMsg = error.message;
                this.doRefresh = false;
              });
      },

      default() {
        return [];
      },
      watch() {
        this.doRefresh
      }
    }
  },
  computed: {
    ...mapGetters(["isLoading", "busStops"]),
    Constants () {
      return Constants;
    }
  },

  methods: {
    ...mapActions({
      "getAllStops": GET_ALL_STOPS
    }),

    onRefresh() {
      this.errorMsg = "";
      this.doRefresh = true;
    }
  }
};
</script>

<style>
.v-progress-linear {
  border-radius: 10px;
}
</style>


