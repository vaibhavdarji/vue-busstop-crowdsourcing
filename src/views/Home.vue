<template>
  <bus-stops></bus-stops>
</template>

<script>
  import BusStops from '@/components/BusStops';
  export default {
    components: {
      BusStops
    }

  }
</script>
