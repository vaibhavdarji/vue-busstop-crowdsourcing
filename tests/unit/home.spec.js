import { shallowMount, mount } from "@vue/test-utils";
import Home from "@/views/Home.vue";
import BusStops from '@/components/BusStops.vue';
import Constants from '@/constants';
import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";

Vue.use(Vuex);
Vue.use(Vuetify);

describe("Home.vue", () => {
  let  getters, store;

  beforeEach(() => {
    getters = {
      isLoading: () => false,
      busStops: () => []
    };
    store = new Vuex.Store({
      getters
    });
  });

  it("should match snapshot", () => {
    const wrapper = shallowMount(Home, {
      Vue
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should contain BusStops.vue as a child component", () => {
    const wrapper = mount(Home, {
      store,
      Vue
    });
    expect(wrapper.contains(BusStops)).toBe(true);
  });

  it("BusStop.vue child component id match", () => {
    const wrapper = mount(Home, {
      store,
      Vue
    });

    const {
      BS_LIST_ID
    } = Constants.APP_TITLE;
    expect(wrapper.find(BusStops).attributes('id')).toBe(BS_LIST_ID);
  });
});