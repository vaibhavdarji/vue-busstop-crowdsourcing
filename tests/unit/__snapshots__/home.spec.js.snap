// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue should match snapshot 1`] = `<bus-stops-stub />`;
